src/Castle.Core/DynamicProxy/ProxyUtil.cs

```diff
@@ -26,11 +26,84 @@ namespace Castle.DynamicProxy
 #endif
 
 	using Castle.Core.Internal;
+	using Castle.DynamicProxy.Generators;
+	using Castle.DynamicProxy.Internal;
 
 	public static class ProxyUtil
 	{
 		private static readonly SynchronizedDictionary<Assembly, bool> internalsVisibleToDynamicProxy = new SynchronizedDictionary<Assembly, bool>();
 
+		/// <summary>
+		///   Attempts to create a delegate of the specified type to a suitable `Invoke` method
+		///   on the given <paramref name="proxy"/> instance.
+		/// </summary>
+		/// <param name="proxy">The proxy instance to which the delegate should be bound.</param>
+		/// <typeparam name="TDelegate">The type of delegate that should be created.</typeparam>
+		/// <param name="delegate">The produced delegate if the method succeeds; otherwise <see langword="null"/>.</param>
+		/// <returns><see langword="true"/> if the method succeeds; otherwise <see langword="false"/>.</returns>
+		public static bool TryCreateDelegateToMixin<TDelegate>(object proxy, out TDelegate @delegate)
+		{
+			if (TryCreateDelegateToMixin(proxy, typeof(TDelegate), out var d))
+			{
+				@delegate = (TDelegate)(object)d;
+				return true;
+			}
+			else
+			{
+				@delegate = default(TDelegate);
+				return false;
+			}
+		}
+
+		/// <summary>
+		///   Attempts to create a delegate of the specified type to a suitable `Invoke` method
+		///   on the given <paramref name="proxy"/> instance.
+		/// </summary>
+		/// <param name="proxy">The proxy instance to which the delegate should be bound.</param>
+		/// <param name="delegateType">The type of delegate that should be created.</param>
+		/// <param name="delegate">The produced delegate if the method succeeds; otherwise <see langword="null"/>.</param>
+		/// <returns><see langword="true"/> if the method succeeds; otherwise <see langword="false"/>.</returns>
+		public static bool TryCreateDelegateToMixin(object proxy, Type delegateType, out Delegate @delegate)
+		{
+			if (proxy == null)
+			{
+				throw new ArgumentNullException(nameof(proxy));
+			}
+
+			if (delegateType == null)
+			{
+				throw new ArgumentNullException(nameof(delegateType));
+			}
+
+			if (!delegateType.IsDelegateType())
+			{
+				throw new ArgumentException("Type is not a delegate type.", nameof(delegateType));
+			}
+
+			var invokeMethod = delegateType.GetMethod("Invoke");
+			var proxiedInvokeMethod =
+				proxy
+				.GetType()
+				.GetMember("Invoke", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance)
+				.Cast<MethodInfo>()
```

Comment:
Does `GetMember` return null or throw, if returns null won't `Cast<>` throw?

Reply:
Neither: `GetMember` returns an empty array if no matching members are found.

From the [API reference documentation for `Type.GetMember(string, BindingFlags)`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getmember?view=netframework-4.7.2#System_Type_GetMember_System_String_System_Reflection_BindingFlags_):

> #### Returns
> MemberInfo[]
>
> An array of MemberInfo objects representing the public members with the specified name, if found; otherwise, an empty array.

Note also that this documentation incorrectly claims that only public members are returned; if you specify `BindingFlags.NonPublic`, it will also return non-public members. But the documentation *is* correct with respect to returning an empty array in the failure case.

Therefore I'd say that LINQ query should work as intended.